app_history.d.ts

```diff
@@ -0,0 +1,99 @@
+interface Window {
+  readonly appHistory: AppHistory;
+}
+
+interface AppHistory extends EventTarget {
+  readonly current: AppHistoryEntry|null;
+  readonly transition: AppHistoryTransition|null;
+
+  entries(): AppHistoryEntry[];
+
+  readonly canGoBack: boolean;
+  readonly canGoForward: boolean;
+
+  navigate(url: string, options?: AppHistoryNavigateOptions): Promise<void>;
+  reload(options?: AppHistoryReloadOptions): Promise<void>;
+
+  goTo(key: string, options?: AppHistoryNavigationOptions): Promise<void>;
+  back(options?: AppHistoryNavigationOptions): Promise<void>;
+  forward(options?: AppHistoryNavigationOptions): Promise<void>;
+
+  onnavigate: ((ev: Event) => any)|null;
+  onnavigatesuccess: ((ev: Event) => any)|null;
+  onnavigateerror: ((ev: Event) => any)|null;
+}
+
+interface AppHistoryTransition {
+  readonly navigationType: AppHistoryNavigationType;
+  readonly from: AppHistoryEntry;
+  readonly finished: Promise<void>;
+
+  rollback(options?: AppHistoryNavigationOptions): Promise<void>;
+}
+
+interface AppHistoryEntry extends EventTarget {
+  readonly key: string;
+  readonly id: string;
+  readonly url: string;
+  readonly index: number;
+  readonly sameDocument: boolean;
+
+  getState(): unknown;
+
+  onnavigateto: ((ev: Event) => any)|null;
+  onnavigatefrom: ((ev: Event) => any)|null;
+  onfinish: ((ev: Event) => any)|null;
+  ondispose: ((ev: Event) => any)|null;
+}
+
+type AppHistoryNavigationType = 'reload'|'push'|'replace'|'traverse';
+
+interface AppHistoryNavigationOptions {
+  navigateInfo?: unknown;
+}
+
+interface AppHistoryNavigateOptions extends AppHistoryNavigationOptions {
+  state?: unknown;
+  replace?: boolean;
+}
+
+interface AppHistoryReloadOptions extends AppHistoryNavigationOptions {
+  state?: unknown;
+}
+
+declare class AppHistoryNavigateEvent extends Event {
+  constructor(type: string, eventInit?: AppHistoryNavigateEventInit);
+
+  readonly navigationType: AppHistoryNavigationType;
+  readonly canRespond: boolean;
+  readonly userInitiated: boolean;
+  readonly hashChange: boolean;
+  readonly destination: AppHistoryDestination;
+  readonly signal: AbortSignal;
+  readonly formData: FormData|null;
+  readonly info: unknown;
+
+  respondWith(newNavigationAction: Promise<void>): void;
+}
+
+interface AppHistoryNavigateEventInit extends EventInit {
+  navigationType?: AppHistoryNavigationType;
+  canRespond?: boolean;
+  userInitiated?: boolean;
+  hashChange?: boolean;
+  destination?: AppHistoryDestination;
```

Comment:
I could be wrong, but is a destination just an `AppHistoryEntry`?

Reply:
No; see #139.

Reply:
Nit: destination is not optional in the IDL. (I.e. you cannot do `new AppHistoryNavigateEvent("x", {})`, you need to include `{ destination: somePreexistingDestination }`.)

Same for `signal`.

Reply:
I discovered by accidental earlier today that the impl in Canary does treat destination as nullable. Does that merit a Chromium issue or is it just a WIP artifact?

Reply:
It looks like the issue is with both the spec and implementation in that they mark the init argument to the constructor as optional, when they should not. I'll fix the spec and file an issue.